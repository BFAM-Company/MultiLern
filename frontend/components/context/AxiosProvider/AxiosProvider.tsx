import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";
import AsyncStorage from "@react-native-async-storage/async-storage";
// import * as Keychain from "react-native-keychain";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance
}

const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    baseURL: "http://localhost:3001/auth",
  }),
  publicAxios: axios.create({
    baseURL: "http://localhost:3001/",
  })
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    baseURL: "http://localhost:3001/auth",
  });

  const publicAxios = axios.create({
    baseURL: "http://localhost:3001/",
  });

  authAxios.interceptors.request.use(
    (config) => {
      if (!config.headers.Authorization) {
        config.headers.Authorization = `Bearer ${authContext?.authState.accessToken}`;
      }

      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    const options = {
      method: "POST",
      data,
      url: "http://localhost:3001/auth/refresh",
    };

    return axios(options)
      .then(async (tokenRefreshResponse: any) => {
        failedRequest.response.config.headers.Authorization =
          "Bearer " + tokenRefreshResponse.data;

        await AsyncStorage.setItem('auth',JSON.stringify({
            accessToken: tokenRefreshResponse.data,
            refreshToken: authContext?.authState.refreshToken,
        }))

        authContext?.setAuthState({
          ...authContext.authState,
          accessToken: tokenRefreshResponse.data,
        });
        
        return Promise.resolve();
      })
      .catch(async (e: any) => {
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false
        });
        await AsyncStorage.clear()
      });
  };

  createAuthRefreshInterceptor(authAxios, refreshAuthLogic, {});

  return (
    <Provider
      value={{
        authAxios,
        publicAxios,
      }}
    >
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
