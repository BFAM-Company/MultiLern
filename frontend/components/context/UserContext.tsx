import React, { createContext, useState, ReactNode } from "react";

export interface UserDataState {
  nickname: string;
  email: string;
  avatar: any;
  isLogged: boolean
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}


const UserDataContext = createContext<UserDataContextType | null>(null);
const { Provider } = UserDataContext;

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState>();

  return (
    <Provider
      value={{
        userData,
        setUserData
      }}
    >
      {children}
    </Provider>
  );
};

export { UserDataContext, UserDataProvider };
