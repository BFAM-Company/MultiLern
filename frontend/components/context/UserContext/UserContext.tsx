import React, { createContext, useState, ReactNode } from "react";

export interface UserDataState {
  id: number,
  nickname: string;
  email: string;
  avatar: any;
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}


const UserDataContext = createContext<UserDataContextType | null>(null);
const { Provider } = UserDataContext;

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState>();

  return (
    <Provider
      value={{
        userData,
        setUserData
      }}
    >
      {children}
    </Provider>
  );
};

export { UserDataContext, UserDataProvider };
