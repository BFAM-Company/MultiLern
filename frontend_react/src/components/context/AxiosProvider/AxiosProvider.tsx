import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance;
}

// Kontekst Axios
const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  }),
  publicAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  }),
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  });

  const publicAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  });

  // Interceptor dodający nagłówki autoryzacji
  authAxios.interceptors.request.use(
    (config) => {
      if (authContext?.authState.accessToken) {
        config.headers.Authorization = `Bearer ${authContext.authState.accessToken}`;
      }
      return config;
    },
    (error) => Promise.reject(error)
  );

  // Funkcja obsługująca odświeżanie tokenu
  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    return axios
      // .post(`${process.env.REACT_APP_API_URL}/auth/refresh`, data)
      .post(`${import.meta.env.VITE_API_URL}/auth/refresh`, data)
      .then((tokenRefreshResponse) => {
        const newAccessToken = tokenRefreshResponse.data;

        // Ustawienie nowego tokenu w localStorage
        localStorage.setItem(
          "auth",
          JSON.stringify({
            accessToken: newAccessToken,
            refreshToken: authContext?.authState.refreshToken,
          })
        );

        // Aktualizacja stanu kontekstu
        authContext?.setAuthState((prevState) => ({
          ...prevState,
          accessToken: newAccessToken,
        }));

        // Ponowienie oryginalnego żądania
        failedRequest.response.config.headers.Authorization = `Bearer ${newAccessToken}`;
        return Promise.resolve();
      })
      .catch(() => {
        // Wylogowanie w przypadku niepowodzenia
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false,
        });
        localStorage.clear();
      });
  };

  // Dodanie interceptora obsługującego odświeżanie tokenu
  createAuthRefreshInterceptor(authAxios, refreshAuthLogic);

  return (
    <Provider value={{ authAxios, publicAxios }}>
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
