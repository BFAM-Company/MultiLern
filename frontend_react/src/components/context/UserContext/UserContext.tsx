import React, { createContext, useState, ReactNode } from "react";

// Typy danych uÅ¼ytkownika
export interface UserDataState {
  id: number;
  nickname: string;
  email: string;
  avatar: string | null;
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}

// Tworzenie kontekstu
const UserDataContext = createContext<UserDataContextType | null>(null);

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState | undefined>(undefined);

  return (
    <UserDataContext.Provider
      value={{
        userData,
        setUserData,
      }}
    >
      {children}
    </UserDataContext.Provider>
  );
};

export { UserDataContext, UserDataProvider };
