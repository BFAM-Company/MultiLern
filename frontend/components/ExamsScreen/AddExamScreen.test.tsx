import React from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react-native";
import AddExamScreen from "./AddExamScreen";
import { AxiosContext } from '../context/AxiosProvider/AxiosProvider';
import { UserDataContext } from '../context/UserContext/UserContext';
import * as ImagePicker from "expo-image-picker";
import axios from "axios";

jest.mock("@react-native-async-storage/async-storage", () => require("../../__mocks__/mock-async-storage"));
jest.useFakeTimers();

describe('AddExamScreen', () => {
    const mockPageSwitcher = jest.fn();
    const mockUserContext = {
        setUserData: jest.fn(),
        userData: {
            id: 1,
            nickname: 'JohnDoe',
            avatar: 'https://example.com/avatar.jpg',
            email: 'john.doe@example.com',
        }
    };
    const mockAxiosContext = {
        publicAxios: {
            post: jest.fn()
        },
        authAxios: {}
    };

    const mockPublicAxios = axios.create();
    const mockAuthAxios = axios.create();

    const setup = () => {
        return render(
            <UserDataContext.Provider value={mockUserContext}>
                <AxiosContext.Provider value={{ authAxios: mockPublicAxios, publicAxios: mockPublicAxios }}>
                    <AddExamScreen pageSwitcher={mockPageSwitcher} />
                </AxiosContext.Provider>
            </UserDataContext.Provider>
        );
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { toJSON } = setup();
        expect(toJSON()).toMatchSnapshot();
    });

    it('handles text input correctly', () => {
        const { getByPlaceholderText } = setup();
        const titleInput = getByPlaceholderText('tytuł, dział lub zagadnienie...');
        const contentInput = getByPlaceholderText('Treści zadań...');

        fireEvent.changeText(titleInput, 'Test Title');
        fireEvent.changeText(contentInput, 'Test Content');

        expect(titleInput.props.value).toBe('Test Title');
        expect(contentInput.props.value).toBe('Test Content');
    });

    it('adds and removes images correctly', async () => {
        const { getByTestId, getByText } = setup();

        jest.spyOn(ImagePicker, 'requestMediaLibraryPermissionsAsync').mockResolvedValueOnce({
            status: ImagePicker.PermissionStatus.GRANTED,
            expires: 'never',
            granted: true,
            canAskAgain: true,
        });
        jest.spyOn(ImagePicker, 'launchImageLibraryAsync').mockResolvedValueOnce({
            canceled: false,
            assets: [{
                base64: "testbase64",
                uri: "",
                width: 0,
                height: 0
            }]
        });

        const uploadButton = getByTestId('uploadButton');
        await act(async () => {
            fireEvent.press(uploadButton);
        });

        await waitFor(() => {
            expect(getByTestId('image-0')).toBeTruthy();
        });

        const addedImage = getByTestId('image-0');
        fireEvent.press(addedImage);

        await waitFor(() => {
            expect(() => getByTestId('image-0')).toThrow();
        });
    });

    it('submits the exam correctly', async () => {
        const { getByText, getByPlaceholderText } = setup();

        const titleInput = getByPlaceholderText('tytuł, dział lub zagadnienie...');
        const contentInput = getByPlaceholderText('Treści zadań...');
        fireEvent.changeText(titleInput, 'Test Title');
        fireEvent.changeText(contentInput, 'Test Content');

        const submitButton = getByText('Prześlij');

        mockPublicAxios.post = jest.fn().mockResolvedValueOnce({ status: 200 });

        await act(async () => {
            fireEvent.press(submitButton);
        });

        await waitFor(() => {
            expect(mockPublicAxios.post).toHaveBeenCalledWith(
                '/posts/exam',
                {
                    title: 'Test Title',
                    content: 'Test Content',
                    category: '',
                    date: expect.any(String),
                    images: [],
                    tags: [],
                    userId: 1,
                },
                {
                    headers: { 'Content-Type': 'application/json' }
                }
            );
            expect(mockPageSwitcher).toHaveBeenCalledWith('UserExercises', { userId: 1 });
        });
    });
});
