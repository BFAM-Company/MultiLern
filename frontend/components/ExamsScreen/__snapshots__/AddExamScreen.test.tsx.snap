// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddExamScreen renders correctly 1`] = `
<RCTScrollView
  style={
    {
      "height": "auto",
      "width": 750,
    }
  }
>
  <View>
    <RCTScrollView
      style={
        {
          "width": 750,
        }
      }
    >
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "white",
              "display": "flex",
              "height": "auto",
              "justifyContent": "flex-start",
              "padding": 20,
              "width": 750,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgb(45,45,55)",
                "fontSize": 20,
                "fontWeight": "900",
                "marginBottom": 20,
                "marginTop": 50,
              }
            }
          >
            Dodaj tytuł, dział lub zagadnienie
          </Text>
          <Text
            style={
              {
                "color": "gray",
                "fontSize": 16,
                "marginBottom": 20,
              }
            }
          >
            Ułatwi to późniejsze znalezienie go innym użytkownikom
          </Text>
          <TextInput
            onChangeText={[Function]}
            placeholder="tytuł, dział lub zagadnienie..."
            style={
              [
                {
                  "backgroundColor": "#fff",
                  "borderTopLeftRadius": 10,
                  "borderTopRightRadius": 10,
                  "elevation": 16,
                  "padding": 10,
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0.44,
                  "shadowRadius": 10.32,
                  "width": "90%",
                },
                {
                  "height": 50,
                },
              ]
            }
            value=""
          />
          <Text
            style={
              {
                "color": "rgb(45,45,55)",
                "fontSize": 20,
                "fontWeight": "900",
                "marginBottom": 20,
                "marginTop": 50,
              }
            }
          >
            Dodaj kategorie sprawdzianu
          </Text>
          <View
            style={
              [
                {
                  "alignSelf": "stretch",
                },
                undefined,
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="ios_touchable_wrapper"
            >
              <View
                pointerEvents="box-only"
              >
                <TextInput
                  editable={false}
                  style={
                    [
                      undefined,
                      {
                        "color": "#c7c7cd",
                      },
                    ]
                  }
                  testID="text_input"
                  value="Wybierz kategorie..."
                />
              </View>
            </View>
          </View>
          <Text
            style={
              {
                "color": "rgb(45,45,55)",
                "fontSize": 20,
                "fontWeight": "900",
                "marginBottom": 20,
                "marginTop": 50,
              }
            }
          >
            Dodaj treść pytań
          </Text>
          <Text
            style={
              {
                "color": "gray",
                "fontSize": 16,
                "marginBottom": 20,
              }
            }
          >
            Wpisz treści zadań lub dodaj zdjęcia poniżej
          </Text>
          <TextInput
            editable={true}
            multiline={true}
            onChangeText={[Function]}
            placeholder="Treści zadań..."
            style={
              [
                {
                  "backgroundColor": "#fff",
                  "borderTopLeftRadius": 10,
                  "borderTopRightRadius": 10,
                  "elevation": 16,
                  "padding": 10,
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0.44,
                  "shadowRadius": 10.32,
                  "width": "90%",
                },
                {
                  "minHeight": 400,
                },
              ]
            }
            value=""
          />
          <Text
            style={
              {
                "color": "rgb(45,45,55)",
                "fontSize": 20,
                "fontWeight": "900",
                "marginBottom": 20,
                "marginTop": 50,
              }
            }
          >
            Dodaj zdjęcia, jeżeli takie posiadasz
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "flexWrap": "wrap",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="uploadButton"
            >
              <View
                style={
                  {
                    "borderBlockColor": "black",
                    "borderRadius": 5,
                    "borderWidth": 1,
                    "height": 70,
                    "padding": 10,
                    "width": 70,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    {
                      "height": 50,
                      "opacity": 0.5,
                      "width": 50,
                    }
                  }
                />
              </View>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "borderRadius": 20,
                "height": 55,
                "margin": 20,
                "opacity": 1,
                "width": 300,
              }
            }
          >
            <ViewManagerAdapter_ExpoLinearGradient
              colors={
                [
                  4281150775,
                  4281150775,
                  4284769390,
                ]
              }
              style={
                {
                  "alignItems": "center",
                  "borderRadius": 20,
                  "display": "flex",
                  "height": "100%",
                  "justifyContent": "center",
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 22,
                    "fontWeight": "900",
                  }
                }
              >
                Prześlij
              </Text>
            </ViewManagerAdapter_ExpoLinearGradient>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTScrollView>
`;
